### PR TITLE
Add cross-compilation and testing to Arm64 on CI

### PR DESCRIPTION
**Pull request by hug-dev**
Modifies the code so that the crate only writes to the OUT_DIR directory
so it can be used with cross.
Add cross-compilation and testing on the CI.

### CODE REVIEW

**Comment by hug-dev** [comment]
@ionut-arm I use `Command` to directly call `wget` and `tar`, maybe there is a better way to do this? Because of this, I have to add a `Dockerfile` that installs `wget` 💣

File: build.rs [hunk 1]
```diff
@@ -57,29 +66,49 @@ fn generate_proto_sources() -> Result<()> {
...
+fn get_protobuf_files() {
     // TODO: Use semantic versioning to get the newest versions.
-    Command::new("git")
-        .arg("clone")
-        .arg("https://github.com/parallaxsecond/parsec-operations.git")
-        .arg("-b")
-        .arg(PARSEC_OPERATIONS_VERSION)
-        .arg("target/parsec-operations")
-        .output()
+    if !Command::new("wget")
```
Comment (ionut-arm) [inline build.rs]: [Curl](https://crates.io/crates/curl) might be easier to use. Using Rust only

File: build.rs [hunk 2]
```diff
@@ -57,29 +66,49 @@ fn generate_proto_sources() -> Result<()> {
...
+        ))
+        .status()
+        .expect("wget command failed.")
+        .success()
+    {
+        panic!("wget command returned an error status.");
+    }
+
+    // Gets extracted as parsec-operations-PARSEC_OPERATIONS_VERSION directory.
+    if !Command::new("tar")
```
Comment (ionut-arm) [inline build.rs]: You have bindings to [tar](https://crates.io/crates/tar) as a library too.

File: src/operations_protobuf/mod.rs [hunk 3]
```diff
@@ -26,8 +26,28 @@ mod convert_asym_verify;
 mod convert_list_providers;
 mod convert_list_opcodes;
 
+// Include the Rust generated file in its own module.
+macro_rules! include_protobuf_as_module {
```
Comment (ionut-arm) [inline src/operations_protobuf/mod.rs]: Nice 😎

File: build.rs [hunk 4]
```diff
@@ -12,20 +12,24 @@
...
-use std::process::{Command, Output};
+use std::process::Command;
 
-const PROTO_FOLDER: &str = "target/parsec-operations/protobuf";
-const PROTO_OUT_DIR: &str = "src/operations_protobuf/generated_ops";
 const PARSEC_OPERATIONS_VERSION: &str = "0.2.0";
 
 // TODO: handle OsStrings more carefully, as .into_string() might fail
 
 fn generate_mod_file() -> Result<()> {
```
Comment (ionut-arm) [inline build.rs]: I don't think you need this file anymore, you've replaced it with a fully-contained module in `src/operations_protobuf/mod.rs`

File: build.rs (hunk 1 above)
Comment (ionut-arm) [reply build.rs]: Also, if you manage to use curl, you don't need the whole shenanigans of installing wget/having a special Docker image.

File: src/operations_protobuf/mod.rs (hunk 3 above)
Comment (ionut-arm) [reply src/operations_protobuf/mod.rs]: You can move this macro definition inside the `generated_ops` module, don't think you'd be using it anywhere else, right?

**Comment by hug-dev** [comment]
I did not modify the direct command calls of `wget` and `tar` as I was facing the following error when trying to use the `clang` crate in the build script:
```
  = note: /usr/bin/ld: /target/debug/deps/libopenssl_sys-07dcfde38d575d2d.rlib(s3_clnt.o): Relocations in generic ELF (EM: 183)
          /usr/bin/ld: /target/debug/deps/libopenssl_sys-07dcfde38d575d2d.rlib(s3_clnt.o): Relocations in generic ELF (EM: 183)
          /target/debug/deps/libopenssl_sys-07dcfde38d575d2d.rlib: error adding symbols: File in wrong format
          collect2: error: ld returned 1 exit status
```
I think there might be a clash between the OpenSSL library on the host and the one used in the Docker container so that linking fails. This will need more investigating.

**Pull request merged**